File: dart/lib/src/piece.dart
enum PieceType { king, queen, rook, bishop, knight, pawn }

enum PieceColor { white, black }

class Piece {
  final PieceType type;
  final PieceColor color;

  Piece(this.type, this.color);

  factory Piece.fromChar(String char) {
    final color = char.toUpperCase() == char ? PieceColor.white : PieceColor.black;
    final type = charToType(char.toLowerCase());
    return Piece(type, color);
  }

  String toChar() {
    final char = _typeToChar(type);
    return color == PieceColor.white ? char.toUpperCase() : char.toLowerCase();
  }

  static PieceType charToType(String char) {
    switch (char) {
      case 'k':
        return PieceType.king;
      case 'q':
        return PieceType.queen;
      case 'r':
        return PieceType.rook;
      case 'b':
        return PieceType.bishop;
      case 'n':
        return PieceType.knight;
      case 'p':
        return PieceType.pawn;
      default:
        throw ArgumentError('Invalid piece character: $char');
    }
  }

  static String _typeToChar(PieceType type) {
    switch (type) {
      case PieceType.king:
        return 'k';
      case PieceType.queen:
        return 'q';
      case PieceType.rook:
        return 'r';
      case PieceType.bishop:
        return 'b';
      case PieceType.knight:
        return 'n';
      case PieceType.pawn:
        return 'p';
    }
  }
}

File: go/internal/game/piece.go
package game

type Piece byte

const (
	Empty Piece = iota
	Pawn
	Knight
	Bishop
	Rook
	Queen
	King
)

type Color byte

const (
	White Color = iota
	Black
)

func (p Piece) Color() Color {
	if p >= Pawn && p <= King {
		return White
	}
	if p >= Pawn+6 && p <= King+6 {
		return Black
	}
	return 2 // No color
}

func FromChar(c rune) Piece {
	switch c {
	case 'P':
		return Pawn
	case 'N':
		return Knight
	case 'B':
		return Bishop
	case 'R':
		return Rook
	case 'Q':
		return Queen
	case 'K':
		return King
	case 'p':
		return Pawn + 6
	case 'n':
		return Knight + 6
	case 'b':
		return Bishop + 6
	case 'r':
		return Rook + 6
	case 'q':
		return Queen + 6
	case 'k':
		return King + 6
	default:
		return Empty
	}
}

func (p Piece) String() string {
	switch p {
	case Pawn:
		return "P"
	case Knight:
		return "N"
	case Bishop:
		return "B"
	case Rook:
		return "R"
	case Queen:
		return "Q"
	case King:
		return "K"
	case Pawn + 6:
		return "p"
	case Knight + 6:
		return "n"
	case Bishop + 6:
		return "b"
	case Rook + 6:
		return "r"
	case Queen + 6:
		return "q"
	case King + 6:
		return "k"
	default:
		return "."
	}
}